sklearn/metrics/pairwise.py

```diff
@@ -161,6 +162,121 @@ def check_paired_arrays(X, Y):
 
 
 # Pairwise distances
+def _euclidean_distances_float64(X, Y, Y_norm_squared=None, squared=False,
+                                 X_norm_squared=None):
+    """
+    Compute the euclidean distances between X and Y when both are arrays of
+    float64.
+    """
+    if X_norm_squared is not None:
+        XX = X_norm_squared
+    else:
+        XX = row_norms(X, squared=True)[:, np.newaxis]
+
+    if X is Y:  # shortcut in the common case euclidean_distances(X, X)
+        YY = XX.T
+    elif Y_norm_squared is not None:
+        YY = Y_norm_squared
+    else:
+        YY = row_norms(Y, squared=True)[np.newaxis, :]
+
+    distances = safe_sparse_dot(X, Y.T, dense_output=True)
+    distances *= -2
+    distances += XX
+    distances += YY
+    np.maximum(distances, 0, out=distances)
+
+    if X is Y:
+        # Ensure that distances between vectors and themselves are set to 0.0.
+        # This may not be the case due to floating point rounding errors.
+        distances.flat[::distances.shape[0] + 1] = 0.0
+
+    return distances if squared else np.sqrt(distances, out=distances)
+
+
+def _euclidean_distances_cast(X, Y, outdtype, Y_norm_squared=None,
+                              squared=False, X_norm_squared=None):
+    """
+    Compute the euclidean distance matrix between X and Y by casting to float64
+    for a greater numerical stability.
+
+    The computation is done by blocks to limit additional memory usage.
+    """
+    # For performance reasons, swap X and Y if I got X_norm_squared but not
+    # Y_norm_squared
+    if X_norm_squared is not None and Y_norm_squared is None:
+        swap = True
+        X, Y = Y, X
+        X_norm_squared, Y_norm_squared = None, X_norm_squared.T
+    else:
+        swap = False
+
+    # No more than 10MB of additional memory will be used to cast X and Y to
+    # float64 and to get the float64 result.
+    maxmem = 10*1024*1024
+    itemsize = np.float64(0).itemsize
+    maxmem /= itemsize
+
+    # Compute the block size that won't use more than maxmem bytes of
+    # additional (temporary) memory.
+    # The amount of additional memory required is:
+    # - a float64 copy of a block of the rows of X if needed;
+    # - a float64 copy of a block of the rows of Y if needed;
+    # - the float64 block result;
+    # - a float64 copy of a block of X_norm_squared if needed;
+    # - a float64 copy of a block of Y_norm_squared if needed.
+    # This is a quadratic equation that we solve to compute the block size that
+    # would use maxmem bytes.
```

Comment:
What happens if we use a bit more? The performance is decreased?

Reply:
The performance decreases a bit, yes. But it's only noticeable for a `maxmem` several order of magnitude larger than required. If you look at the first set of plots, you'll see a trend for larger memory sizes.